File: Python_implementation/receptorCRC.py
from view import *

class ReceptorCRC():

    def __init__(self, response) -> None:

        self.trama = []
        self.res = False
        self.data = response
        for c in response:
            if (c == '1'): self.trama.append(True)
            elif (c == '0'): self.trama.append(False)
        # printTrama(self.trama)

        # generate polinomio
        self.polinomio = [
            True, False, True, True, False, True, False, True, 
            False, False, False, True, False, True, False, True,
            False, False, False, False, True, True, True, False, 
            True, False, False, True, True, False, True, False
        ]

        # printTrama(self.polinomio)

        self.CRC()

    def CRC(self):
        actual = []

        # Initialize actual
        for i in range(len(self.polinomio)):
            actual.append(self.trama[i])

        # print()
        # printTrama(self.trama)
        # printTrama(self.polinomio)
        # printTrama(actual)

        nextB = len(self.polinomio)

        while(nextB < len(self.trama)):

            # remove first 0 and add next value
            actual = actual[1:]
            actual.append(self.trama[nextB])
            nextB += 1

            # check if the first digit is 1
            if (actual[0]):

                # print()
                # printTrama(actual)
                # printTrama(self.polinomio)

                temp = [actual[i] ^ self.polinomio[i] for i in range(len(self.polinomio))]

                # printTrama(temp)

                actual = []
                actual = [x for x in temp]

            # else:
            #     # pasa los 0s
            #     print()
            #     printTrama(actual)

        print("\nResultado receptor")
        printTrama(actual)

        self.residue = actual

        self.cambiado = False

        for b in actual:
            if b: self.cambiado = True

        if (not self.cambiado): 
            print("\nMensaje enviado sin cambios\n")
            self.res = True
        else: 
            print("\nMensaje enviado con errores\n")

    def get_final_results(self):
        if self.res:
            return self.data[:-31]
        else:
            return ""
        
    def detectedError(self):
        if self.cambiado: return 1
        else: return 0
    
    def getResidue(self):

        res = "'"

        for x in self.residue:
            if x:
                res += "1"
            else: 
                res += "0"

        res += "'"

        return res

        

File: Java_implementation/ReceptorCRC.java
import java.util.ArrayList;

public class ReceptorCRC {

    View view = new View();

    ArrayList<Boolean> trama = new ArrayList<>();
    ArrayList<Boolean> result;

    ArrayList<Boolean> polinomio;

    public ReceptorCRC(String tramaS) {

        for (char c: tramaS.toCharArray()) {
            if (c == '1') trama.add(true);
            if (c == '0') trama.add(false);
        }

        // generate polinomio
        polinomio = new ArrayList<>();
        Boolean[] pol = {
            true, false, true, true, false, true, false, true, 
            false, false, false, true, false, true, false, true,
            false, false, false, false, true, true, true, false, 
            true, false, false, true, true, false, true, false
        };
        for (boolean b: pol) polinomio.add(b);

        CRC();

    }


    private void CRC() {

        ArrayList<Boolean> actual = new ArrayList<>();

        // initialize by copying the first bits of the trama
        for (int i = 0; i < polinomio.size(); i++) {
            actual.add(trama.get(i) ^ polinomio.get(i));
        }
        System.out.println();
        view.printTrama(trama);
        view.printTrama(polinomio);
        view.printTrama(actual);

        int nextB = polinomio.size();

        while(nextB < trama.size()) {

            // remove the first 0 and add next bit
            
            actual.remove(0);
            actual.add(trama.get(nextB));
            nextB++;
            
            // check if the first digit is 0            
            if (actual.get(0)) {

                System.out.println();
                view.printTrama(actual);
                view.printTrama(polinomio);

                // oparate normally
                ArrayList<Boolean> temp = new ArrayList<>();
                for (int i = 0; i < polinomio.size(); i ++) {
                    temp.add(actual.get(i) ^ polinomio.get(i));
                }

                view.printTrama(temp);

                // replace
                actual = new ArrayList<>(temp);
            }

            else {
                System.out.println();
                view.printTrama(actual);
            }

        }

        System.out.println("\nResultado receptor");
        view.printTrama(actual);

    }

}
